feat(linter): domains and deps (#4713)
biomejs/biome: crates/biome_analyze/CONTRIBUTING.md

}
```

#### Rule severity

The macro accepts a `severity` field, of type `biome_diagnostics::Severity`. By default, rules without `severity` will start with `Severity::Information`.

If you want to change the default severity, you need to assign it:

```diff
+ use biome_diagnostics::Severity;

declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        language: "js",
        recommended: false,
+       severity: Severity::Warning,
    }
}
```

#### Rule domains

Domains are very specific ways to collect rules that belong to the same "concept". Domains are a way for users to opt-in/opt-out rules that belong to the same domain.

Some examples of domains: testing, specific framework, specific runtime, specific library. A rule can belong to multiple domains.

```diff
+ use biome_analyze::RuleDomain;

declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        language: "js",
        recommended: true,
+       domains: &[RuleDomain::Test],
    }
}
```

Rule domains can unlock various perks in the Biome analyzer:

> - A domain can define a number of `package.json` dependencies. When a user has one or more these dependencies, Biome will automatically enable the rules that belong to the domain. To add/update/remove dependencies to a domain, check the function `RuleDomain::manifest_dependencies`.

Reviewer: ```suggestion
- A domain can define a number of `package.json` dependencies. When a user has one or more of these dependencies, Biome will automatically enable the recommended rules that belong to the domain. To add/update/remove dependencies to a domain, check the function `RuleDomain::manifest_dependencies`.
```

}
```

#### Rule severity

> The macro accepts a `severity` field, which value comes from `biome_diagnostics::Severity`. By default, rules without `severity` will start with `Severity::Information`.

Reviewer: ```suggestion
The macro accepts a `severity` field, of type `biome_diagnostics::Severity`. By default, rules without `severity` will start with `Severity::Information`.
```

If you want to change the default severity, you need to assign it:

```diff
+ use biome_diagnostics::Severity;

declare_lint_rule! {
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        language: "js",
        recommended: false,
+       severity: Severity::Warning,
    }
}
```

#### Rule dependencies

> If you write a rule that are specific to certain libraries, for example `react` or `solid`, you can add them to the metadata.

Reviewer: ```suggestion
If you write a rule that is specific to certain libraries, for example React or Solid, you can add them to the metadata.
```

> Biome will automatically enable those rules by inspecting the dependencies of the closest `package.json`. In the following example, Biome will enable this rule when the dependency `solid` is a dependency:

Reviewer: ```suggestion
Biome will automatically enable those rules by inspecting the dependencies of the closest `package.json`. In the following example, Biome will enable this rule when `solid` is listed in the `dependencies` field:
```
